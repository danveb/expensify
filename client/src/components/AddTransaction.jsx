import { useState, useContext } from "react"; 
import { GlobalContext } from "../context/GlobalState";

const AddTransaction = () => {
    // useState for formData (controlled component for component-level state)
    const [formData, setFormData] = useState({
        text: "", 
        amount: "",
    }); 
    // destructure formData to be used as values 
    const { text, amount } = formData; 

    // useContext
    const { addTransaction } = useContext(GlobalContext); 
    
    // handleChange(e)
    const handleChange = (e) => {
        setFormData((prevState) => ({
            ...prevState,
            [e.target.name]: e.target.value
        }));
    };

    // resetForm
    const resetForm = () => {
        setFormData({
            text: "", 
            amount: "", 
        });
    };

    // handleSubmit 
    const handleSubmit = (e) => {
        e.preventDefault(); 
        const newTransaction = {
            id: Math.floor(Math.random() * 100000000),
            text: text, 
            amount: +amount, // parsing as number
        }; 
        addTransaction(newTransaction); 
        resetForm(); 
    };

    return (
        <>
            <h3>Add new transaction</h3>
            <form onSubmit={handleSubmit}>
                <div className="form-control">
                    <label htmlFor="text">Activity</label>
                    <input 
                        id="text"
                        type="text"
                        name="text"
                        value={text}
                        onChange={handleChange}
                        placeholder="Enter activity..."
                    />
                </div>
                <div className="form-control">
                    <label htmlFor="amount">Amount (negative - expense, positive - income)</label>
                    <input 
                        id="amount"
                        type="number"
                        name="amount"
                        value={amount}
                        onChange={handleChange}
                        placeholder="Enter amount..."
                    />
                </div>
                <button className="btn">Submit</button>
            </form>
        </>
    )
}

export default AddTransaction