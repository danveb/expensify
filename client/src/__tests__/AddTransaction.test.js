import { fireEvent, render } from "@testing-library/react";
import AddTransaction from "../components/AddTransaction"; 

describe("AddTransaction component", () => {
    // smoke test
    test("renders AddTransaction component", () => {
        render(<AddTransaction />); 
    });

    // snapshot test
    test("matches snapshot", () => {
        const { asFragment } = render(<AddTransaction />); 
        expect(asFragment()).toMatchSnapshot(); 
    });

    // getByRole 
    test("should display submit button", () => {
        const { getByRole } = render(<AddTransaction />); 
        const btn = getByRole("button", { name: "Submit" }); 
        expect(btn).toBeInTheDocument(); 
    }); 

    // getByText
    test("should display title of component", () => {
        const { getByText } = render(<AddTransaction />);
        const text = getByText("Add new transaction"); 
        expect(text).toBeInTheDocument(); 
    });

    // getByText
    test("should not dipslay incorrect title of component", () => {
        const { getByText } = render(<AddTransaction />); 
        const text = getByText("Add new transaction"); 
        expect(text).not.toBe("Add New Transaction"); 
    });

    // getByLabelText
    test("should display correct label for activity/amount", () => {
        const { getByLabelText } = render(<AddTransaction />); 
        const activityLabel = getByLabelText("Activity"); 
        const amountLabel = getByLabelText("Amount", { exact: false }); 
        expect(activityLabel).toBeInTheDocument(); 
        expect(amountLabel).toBeInTheDocument(); 
    });

    // getByLabelText
    test("displays initial state for username/password", () => {
        const { getByLabelText } = render(<AddTransaction />);
        const activityInput = getByLabelText("Activity"); 
        const amountInput = getByLabelText("Amount", { exact: false }); 
        expect(activityInput.value).toBe(""); 
        expect(amountInput.value).toBe(""); 
    });

    // getByLabelText / fireEvent
    test("displays activity and amount when added to inputs", () => {
        const { getByLabelText } = render(<AddTransaction />);
        const activity = "Book"; 
        const amount = "30"; 
        const activityInput = getByLabelText("Activity"); 
        const amountInput = getByLabelText("Amount", { exact: false }); 
        fireEvent.change(activityInput, { target: { value: activity }});
        fireEvent.change(amountInput, { target: { value: amount }});
        expect(activityInput.value).toBe("Book"); 
        expect(amountInput.value).toBe("30"); 
    });
});