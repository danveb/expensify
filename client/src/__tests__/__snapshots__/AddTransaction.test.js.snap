// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTransaction component matches snapshot 1`] = `
<DocumentFragment>
  <h3>
    Add new transaction
  </h3>
  <form>
    <div
      class="form-control"
    >
      <label
        for="text"
      >
        Activity
      </label>
      <input
        id="text"
        name="text"
        placeholder="Enter activity..."
        type="text"
        value=""
      />
    </div>
    <div
      class="form-control"
    >
      <label
        for="amount"
      >
        Amount (negative - expense, positive - income)
      </label>
      <input
        id="amount"
        name="amount"
        placeholder="Enter amount..."
        type="number"
        value=""
      />
    </div>
    <button
      class="btn"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
